r/src/RecordBatch.cpp

```diff
@@ -113,6 +113,34 @@ RawVector RecordBatch__to_stream(const std::shared_ptr<arrow::RecordBatch>& batc
   return res;
 }
 
+// [[Rcpp::export]]
+List read_record_batch_stream_(SEXP stream) {
+  RawVector vec(stream);
+
+  std::shared_ptr<arrow::Buffer> buffer;
+  buffer.reset(new arrow::Buffer(vec.begin(), vec.size()));
+
+  io::BufferReader buffer_reader(buffer);
+
+  std::shared_ptr<RecordBatchReader> batch_reader;
+  R_ERROR_NOT_OK(
+      arrow::ipc::RecordBatchStreamReader::Open(&buffer_reader, &batch_reader));
+
+  List batches;
+
+  std::shared_ptr<RecordBatch> chunk;
+  while (true) {
+    R_ERROR_NOT_OK(batch_reader->ReadNext(&chunk));
+    if (chunk == nullptr) {
+      break;
+    }
+
+    batches.push_back(RecordBatch__to_dataframe(chunk));
```

Comment:
😬 you don't really want to use `List::push_back`

Reply:
Looks like `Rcpp::List` is implemented as an `Rcpp::Vector`. However, when I look at the implementation of `push_back()` [Vector.h#L622-L646](https://github.com/RcppCore/Rcpp/blob/f3c5a34e06e774532227470b01c63a8f08ce4313/inst/include/Rcpp/vector/Vector.h#L622-L646), looks like the vector would contain `SEXP` elements. From what I understand `SEXP` would wrap the data frame `List` over each entry in the `Vector` which would cause the `SEXP` to be copied many times, but not the underlined data frame, correct?

The contents inside a batch are usually pretty big so there should be only a very few batches coming from Spark, not sure how other clients would use batches though, so as long as the `SEXP` is not triggering a a copy of the entire data frame for every call to `push_bac()`, which I believe is not, we would only incur the cost of copying a few `SXEP`s which shouldn't be terrible.

That said, if there are use cases where one would create a lot of small batches of `record_batch` records, I agree using `push_back()` here could be an issue. Is there a more efficient way to append an element to a list in `Rcpp`? I also wonder why `Rcpp` relies on this loop:

```cpp
for( ; it < position; ++it, ++target_it){
  *target_it = *it ;
}
```

Instead of `std::copy()` the `SEXP`s, there is probably a good reason I'm missing.

Did I missed something else to consider?